test/condition.test.js

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2019 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+/* eslint-env mocha */
+
+const assert = require('assert');
+const fs = require('fs-extra');
+const nock = require('nock');
+const path = require('path');
+const request = require('request-promise');
+const YAML = require('yaml');
+const Condition = require('../src/Condition.js');
+
+const SPEC_ROOT = path.resolve(__dirname, 'specs/conditions');
+
+describe('Condition tests', () => {
+  fs.readdirSync(SPEC_ROOT).forEach((filename) => {
+    const source = fs.readFileSync(path.resolve(SPEC_ROOT, filename), 'utf8');
+    const document = YAML.parseDocument(source, {
+      merge: true,
+      schema: 'core',
+    });
+    const cfg = document.toJSON() || {};
+    it(`Testing ${filename}`, async () => {
+      try {
+        const cond = new Condition(cfg.condition);
+        if (cfg.vcl) {
+          const vcl = cond.toVCL();
+          assert.equal(vcl, cfg.vcl);
+          assert.equal(null, cfg.error);
+        }
+        nock('http://www.example.com')
+          .get(() => true)
+          .reply(function intercept() {
+            const fn = cond.toFunction();
+            return [200, `${fn(this.req)}`];
+          });
+        const response = await request('http://www.example.com/index.html?a=7');
```

Comment:
why testing a request? wouldn't it be simpler just to pass: `req = { url: "http://www.example.com/index.html?a=7" }` ?

Reply:
The issue says that the function returned by `toFunction` will evaluate the condition against an `express-style` request. There's no `url` in express as far as I can tell:

https://expressjs.com/en/api.html#req

The idea of testing a complete request is to propagate other fields/methods to the `req` object that may be required for evaluating the condition.

Reply:
hmm, ok :-) I think a defined "request" object, that can be derived from express.request by clients would be better. I don't think that we should really depend on express here.

---

I just realized that the `url` is actually not so simple, since the request doesn't include the host. so the VCL needs to be smart and concatenate with the `scheme` and `host` header.

also, the VCL `req.url` doesn't contain the host: https://docs.fastly.com/vcl/variables/req-url/

so I wonder, if we shouldn't remove the `url` as property, and only support `url.host`, `url.path`, etc...

Reply:
I'd like to leave `url` as a shortcut for `and(url.path, url.host)` because it is a common use case and increases the readability. That full URLs don't exist in express or VCL is fine, but shouldn't bother the config writer.

Reply:
> hmm, ok :-) I think a defined "request" object, that can be derived from express.request by clients would be better. I don't think that we should really depend on express here.

So, you suggest to create a mock-up request _manually_ based on the [express API](https://expressjs.com/en/api.html)?

Reply:
> So, you suggest to create a mock-up request _manually_ based on the [express API](https://expressjs.com/en/api.html)?

naah. i think it's ok for now. as long we don't have an explicit express dependency.